modules/imgproc/src/contours_link.cpp

```diff
@@ -0,0 +1,417 @@
+                    {
+                        rns[lower_run].link = rns[lower_run].next;
+                        // First point of contour
+                        ext_rns.push_back(lower_run);
+                    }
+                    n++;
+                    lower_run = rns[rns[lower_run].next].next;
+                }
+                break;
+            case ICV_CONNECTING_ABOVE:
+                if (rns[upper_run].pt.x > rns[rns[lower_run].next].pt.x + 1)
+                {
+                    rns[prev_point].link = rns[lower_run].next;
+                    connect_flag = ICV_SINGLE;
+                    n++;
+                    lower_run = rns[rns[lower_run].next].next;
+                }
+                else
+                {
+                    rns[prev_point].link = upper_run;
+                    if (rns[rns[upper_run].next].pt.x < rns[rns[lower_run].next].pt.x)
+                    {
+                        k++;
+                        prev_point = rns[upper_run].next;
+                        upper_run = rns[rns[upper_run].next].next;
+                    }
+                    else
+                    {
+                        connect_flag = ICV_CONNECTING_BELOW;
+                        prev_point = rns[lower_run].next;
+                        n++;
+                        lower_run = rns[rns[lower_run].next].next;
+                    }
+                }
+                break;
+            case ICV_CONNECTING_BELOW:
+                if (rns[lower_run].pt.x > rns[rns[upper_run].next].pt.x + 1)
+                {
+                    rns[rns[upper_run].next].link = prev_point;
+                    connect_flag = ICV_SINGLE;
+                    k++;
+                    upper_run = rns[rns[upper_run].next].next;
+                }
+                else
+                {
+                    // First point of contour
+                    int_rns.push_back(lower_run);
+
+                    rns[lower_run].link = prev_point;
+                    if (rns[rns[lower_run].next].pt.x < rns[rns[upper_run].next].pt.x)
+                    {
+                        n++;
+                        prev_point = rns[lower_run].next;
+                        lower_run = rns[rns[lower_run].next].next;
+                    }
+                    else
+                    {
+                        connect_flag = ICV_CONNECTING_ABOVE;
+                        k++;
+                        prev_point = rns[upper_run].next;
+                        upper_run = rns[rns[upper_run].next].next;
+                    }
+                }
+                break;
+        }
+    }  // k, n
+
+    for (; n < lower_total / 2; n++)
+    {
+        if (connect_flag != ICV_SINGLE)
+        {
+            rns[prev_point].link = rns[lower_run].next;
+            connect_flag = ICV_SINGLE;
+            lower_run = rns[rns[lower_run].next].next;
+            continue;
+        }
+        rns[rns[lower_run].next] = rns[rns[lower_run].next];
```

Comment:
@mshabunin , why assign it to itself? Is it a bug?

Reply:
Yes, this is a problem, but not serious - there should be no operation, like in the original code:

https://github.com/opencv/opencv/blob/f4ebf7c0a6b2ee845a395ae3dac59927eb862894/modules/imgproc/src/contours.cpp#L1636-L1664

I'll fix it soon.